### PR TITLE
[SPARK-45110][BUILD] Upgrade rocksdbjni to 8.5.3

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr aims to upgrade rocksdbjni from 8.3.2 to 8.5.3.

### Why are the changes needed?
1.The full release notes: 
- https://github.com/facebook/rocksdb/releases/tag/v8.5.3
- https://github.com/facebook/rocksdb/releases/tag/v8.4.4
- https://github.com/facebook/rocksdb/releases/tag/v8.3.3

2.Bug Fixes:
image
- Fix a bug where if there is an error reading from offset 0 of a file from L1+ and that the file is not the first file in the sorted run, data can be lost in compaction and read/scan can return incorrect results.
- Fix a bug where iterator may return incorrect result for DeleteRange() users if there was an error reading from a file.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Manually test:
```
./build/mvn clean install -pl core -am -Dtest.exclude.tags=org.apache.spark.tags.ExtendedLevelDBTest -fn
...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 4.0.0-SNAPSHOT:
[INFO]
[INFO] Spark Project Parent POM ........................... SUCCESS [  7.121 s]
[INFO] Spark Project Tags ................................. SUCCESS [ 10.181 s]
[INFO] Spark Project Local DB ............................. SUCCESS [ 21.153 s]
[INFO] Spark Project Common Utils ......................... SUCCESS [ 14.960 s]
[INFO] Spark Project Networking ........................... SUCCESS [01:01 min]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [ 16.992 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 14.967 s]
[INFO] Spark Project Launcher ............................. SUCCESS [ 11.737 s]
[INFO] Spark Project Core ................................. SUCCESS [38:06 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  40:45 min
[INFO] Finished at: 2023-09-10T17:25:26+08:00
[INFO] ------------------------------------------------------------------------

```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
Let me add the results of org.apache.spark.sql.execution.benchmark.StateStoreBasicOperationsBenchmark.
JDK8: https://github.com/panbingkun/spark/actions/runs/6136869653
JDK11: https://github.com/panbingkun/spark/actions/runs/6136872945
JDK17: https://github.com/panbingkun/spark/actions/runs/6136874414

**Comment by LuciferYang** [comment]
image

Are these bug fixes related to 8.3.2? Spark doesn’t use 8.5.1, right?

**Comment by panbingkun** [comment]
image
The bug fixes listed above are release notes from version 8.3.3, which I believe are bug fixes relative to versions before 8.3.2.

**Comment by HyukjinKwon** [comment]
cc @xuanyuanking and @viirya FYI

**Comment by LuciferYang** [comment]
also cc @dongjoon-hyun

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 1]
```diff
@@ -225,7 +225,7 @@ parquet-jackson/1.13.1//parquet-jackson-1.13.1.jar
 pickle/1.3//pickle-1.3.jar
 py4j/0.10.9.7//py4j-0.10.9.7.jar
 remotetea-oncrpc/1.1.2//remotetea-oncrpc-1.1.2.jar
-rocksdbjni/8.3.2//rocksdbjni-8.3.2.jar
+rocksdbjni/8.5.3//rocksdbjni-8.5.3.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:228]: Unfortunately, Apache Spark 3.5.0 RC5 is also using RocksDB 8.3.2.

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:228]: Maybe, we need to change the issue type of SPARK-45110. Currently, it's filed as `Improvement` for Spark 4.0.0.

**Review by viirya — COMMENTED** [review]
For master branch, looks okay and CI passed too. But as @dongjoon-hyun said, the issue version is not accurate.

File: dev/deps/spark-deps-hadoop-3-hive-2.3 (hunk 1 above)
Comment (panbingkun) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:228]: Done
image

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you, @panbingkun and all.
Merged to master/3.5 for Apache Spark 3.5.1.

**Comment by dongjoon-hyun** [comment]
It seems that GitHub Action complains with a segfault from this RocksDB version.
- https://github.com/apache/spark/actions/runs/6156730315/job/16706042686

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fc89e1ef7e7, pid=83569, tid=0x00007fc87fffd640
#
# JRE version: OpenJDK Runtime Environment (8.0_382-b05) (build 1.8.0_382-b05)
# Java VM: OpenJDK 64-Bit Server VM (25.382-b05 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [librocksdbjni8661574108371776748.so+0x3ef7e7]  rocksdb::DBImpl::GetProperty(rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, std::string*)+0x57
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/runner/work/spark/spark/sql/core/hs_err_pid83569.log
Compiled method (nm)  246314 30882     n 0       org.rocksdb.RocksDB::getProperty (native)
 total in heap  [0x00007fc8fc7af7d0,0x00007fc8fc7afb60] = 912
 relocation     [0x00007fc8fc7af8f8,0x00007fc8fc7af940] = 72
 main code      [0x00007fc8fc7af940,0x00007fc8fc7afb58] = 536
 oops           [0x00007fc8fc7afb58,0x00007fc8fc7afb60] = 8
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

Let me revert this RocksDB version.

**Comment by viirya** [comment]
+1 for revert it.

**Comment by dongjoon-hyun** [comment]
This is reverted.
- master: https://github.com/apache/spark/commit/f33c2051c01eb09f6bbb602ed3c7c637e1c6f421
- branch-3.5: https://github.com/apache/spark/commit/6a2284feaac4f632d645a93361d29e693eeb9d32

**Comment by LuciferYang** [comment]
@dongjoon-hyun @viirya Actually, I have observed several core dumps in version 8.3.2:

https://issues.apache.org/jira/browse/SPARK-44580

```
#
17177# A fatal error has been detected by the Java Runtime Environment:
17178#
17179#  SIGSEGV (0xb) at pc=0x00007f8a077d2743, pid=4403, tid=0x00007f89cadff640
17180#
17181# JRE version: OpenJDK Runtime Environment (8.0_372-b07) (build 1.8.0_372-b07)
17182# Java VM: OpenJDK 64-Bit Server VM (25.372-b07 mixed mode linux-amd64 compressed oops)
17183# Problematic frame:
17184# C  [librocksdbjni886380103972770161.so+0x3d2743]  rocksdb::DBImpl::FailIfCfHasTs(rocksdb::ColumnFamilyHandle const*) const+0x23
17185#
17186# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
17187#
17188# An error report file with more information is saved as:
17189# /home/runner/work/spark/spark/sql/core/hs_err_pid4403.log
17190#
17191# If you would like to submit a bug report, please visit:
17192#   https://github.com/adoptium/adoptium-support/issues
17193# The crash happened outside the Java Virtual Machine in native code.
17194# See problematic frame for where to report the bug.
17195# 
```

So should we revert to version 8.1.1.1?

**Comment by dongjoon-hyun** [comment]
That's too bad. In the worst case, we might consider that too, but I'm not sure.

- Apache Spark 3.4.1 is using `rocksdbjni-7.9.2.jar`.
- Apache Spark 3.5.0 is shipped with `rocksdbjni-8.3.2.jar` already.

Is 8.1.1.1 stable?

**Comment by LuciferYang** [comment]
At least, I haven't observed any core dump cases in 8.1.1.1. On the other hand, I'm conducting targeted r/w stress tests on version 8.1.1.1. It will take a few days before I can reach further conclusions.

**Comment by viirya** [comment]
FTW, is it possible to get the core dump and submit an issue to rocksdb/rocksdbjni project?

**Comment by LuciferYang** [comment]
```
# C  [librocksdbjni8661574108371776748.so+0x3ef7e7]  rocksdb::DBImpl::GetProperty(rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, std::string*)+0x57
```

I have changed some code and tried to call the getDBProperty method on a closed RocksDB instance, a similar error occurs. This might not be a problem with rocksdb, rather it could be due to some asynchronous operations in the business code leading to accessing a closed instance...

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000015b8d16b0, pid=23605, tid=26371
#
# JRE version: OpenJDK Runtime Environment Zulu17.44+15-CA (17.0.8+7) (build 17.0.8+7-LTS)
# Java VM: OpenJDK 64-Bit Server VM Zulu17.44+15-CA (17.0.8+7-LTS, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [librocksdbjni15748758312520513152.jnilib+0x1056b0]  rocksdb::DBImpl::GetProperty(rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>*)+0x54
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/yangjie01/SourceCode/git/spark-mine-sbt/sql/core/hs_err_pid23605.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azul.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by panbingkun** [comment]
Let me investigate it.

**Comment by dongjoon-hyun** [comment]
Thank you for reproducing it, @LuciferYang !

It sounds like a bug still because SIGSEGV should not happen even in that abuse use case. We expect some Java-layer Exception like `IllegalStateException`
>  This might not be a problem with rocksdb, rather it could be due to some asynchronous operations in the business code leading to accessing a closed rocksdb instance...

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya and @panbingkun , too.

**Comment by LuciferYang** [comment]
> Thank you for reproducing it, @LuciferYang !
> 
> It sounds like a bug still because SIGSEGV should not happen even in that abuse use case. We expect some Java-layer Exception like `IllegalStateException`
> 
> > This might not be a problem with rocksdb, rather it could be due to some asynchronous operations in the business code leading to accessing a closed rocksdb instance...

Filed an issue on rocksdb: https://github.com/facebook/rocksdb/issues/11825

**Comment by panbingkun** [comment]
when I revert version 8.1.1.1, accessing DB while it is closed will still generate the following core messages:

```
15:31:12.496 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001207d0d20, pid=93758, tid=0x0000000000007d03
#
# JRE version: OpenJDK Runtime Environment (Zulu 8.66.0.15-CA-macos-aarch64) (8.0_352-b08) (build 1.8.0_352-b08)
# Java VM: OpenJDK 64-Bit Server VM (25.352-b08 mixed mode bsd-aarch64 compressed oops)
# Problematic frame:
# C  [librocksdbjni6268389328305132469.jnilib+0x24d20]  Java_org_rocksdb_RocksDB_getProperty+0x58
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/panbingkun/Developer/spark/spark-community/sql/core/hs_err_pid93758.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azul.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Warning: Unable to read from client, please check on client for further details of the problem.
```

**Pull request closed without merging**